### PR TITLE
Ch sso

### PR DESCRIPTION
**Pull request by chartjes**
This PR

* [x] adds the ability to use OpenCFP Central as a single-sign-on source

### CODE REVIEW

File: config/development.yml.dist [hunk 1]
```diff
@@ -68,3 +68,12 @@ security:
 reviewer:
   users:
     -
+
+opencfpcentral:
```
Comment (localheinz) [inline config/development.yml.dist]: When first running

```
$ make
```

after checkout, the process failed with

```
  The service "OpenCFP\Http\Action\Security\CentralAction" has a dependency on a non-existent parameter "opencfpcentral.clientId".
```

Maybe we should extract parameters and use [`incenteev/parameter-handler`](https://github.com/Incenteev/ParameterHandler)? This is clearly out of scope of this PR, though.

File: resources/views/security/login.twig [hunk 2]
```diff
@@ -13,10 +13,13 @@
       <input id="email" type="email" name="email" value="{% if email %}{{ email | escape }}{% endif %}" class="w-full" placeholder="you@example.com">
       <label for="password" class="text-sm text-white">Password</label>
       <input id="password" type="password" name="password" class="w-full" placeholder="•••••••">
-      <p class="text-right my-4"><a href="{{ url('forgot_password') }}" class="text-white text-sm">Forgot password?</a></p>
+      <p class="text-right my-4"><a href="{{ url('forgot_password') }}" class="text-white text-sm">Forgot your password? {{ sso }}</a></p>
```
Comment (localheinz) [inline resources/views/security/login.twig]: Not sure, why are we rendering `{{ sso }}` here?

screen shot 2018-10-03 at 09 16 55

File: src/Http/Action/Security/CentralAction.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        string $redirectUri,
+        string $authorizeUrl,
+        string $sso
+    ) {
+        $this->authentication = $authentication;
+        $this->urlGenerator   = $urlGenerator;
+        $this->clientId       = $clientId;
+        $this->redirectUri    = $redirectUri;
+        $this->authorizeUrl   = $authorizeUrl;
+        $this->sso            = $sso;
```
Comment (localheinz) [inline src/Http/Action/Security/CentralAction.php]: This field is actually never used, can we remove it, as well as the dependency on the `$sso` parameter?

File: src/Http/Action/Security/LogInAction.php [hunk 4]
```diff
@@ -30,6 +30,9 @@
      */
     private $twig;
 
+    /** @var string */
+    private $sso;
```
Comment (localheinz) [inline src/Http/Action/Security/LogInAction.php]: This field is never used, can we remove it, or revert all of the changes to this file?

File: resources/config/routing.yml [hunk 5]
```diff
@@ -258,3 +257,13 @@ privacy:
    methods: [GET]
    controller: OpenCFP\Http\Action\Signup\PrivacyAction
 
+central:
+  path: /central
+  methods: [GET]
+  controller: OpenCFP\Http\Action\Security\CentralAction
+
+redirect:
+  path: /redirect
```
Comment (localheinz) [inline resources/config/routing.yml]: These paths seem a bit generic, perhaps it makes sense to rename a few of the paths?

* `/opencfp-central`
* `/opencfp-central/redirect`

or

* `/sso`
* `/sso/redirect`

What do you think?

In the same light, perhaps renaming the actions also makes sense?

File: resources/config/routing.yml [hunk 6]
```diff
@@ -258,3 +257,13 @@ privacy:
    methods: [GET]
    controller: OpenCFP\Http\Action\Signup\PrivacyAction
 
+central:
+  path: /central
```
Comment (localheinz) [inline resources/config/routing.yml]: When navigating to `/central`, after a redirect I end up here:

screen shot 2018-10-03 at 10 24 43

What am I missing here?

Also see the `opencfpcentral.authorizeUrl` parameter.

File: src/Http/Action/Security/RedirectAction.php [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        /**
+         * What do we do with the access token?
+         * Check if the user exists
+         * -> if they don't, generate a random password and then create and activate an account
+         * manually log them in
+         * redirect them to their dashboard
+         */
+        $details = \json_decode((string) $response->getBody(), true);
+
+        $user_response = $this->httpClient->get($this->resourceUri, [
```
Comment (localheinz) [inline src/Http/Action/Security/RedirectAction.php]: Should we use `camelCase` here and elsewhere for variable names?

File: src/Http/Action/Security/ShowLogInAction.php [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+         * Otherwise show the login page
+         */
+        try {
+            $this->authentication->authenticate(
+                $request->get('email'),
+                $request->get('password')
+            );
+            $url = $this->urlGenerator->generate('dashboard');
+
+            return new HttpFoundation\RedirectResponse($url);
```
Comment (localheinz) [inline src/Http/Action/Security/ShowLogInAction.php]: Can we move this and the previous statement to after the `try`-and-`catch` block?

I suppose the only thing that can go wrong here and that we care about is the authentication.

File: config/development.yml.dist (hunk 1 above)
Comment (chartjes) [reply config/development.yml.dist]: Make now passes for everything I've done here (saw similar errors but thanks for pointing out I forgot to run it) and I am open to using that parameter-handler but in a follow-up PR once someone other than me can get SSO up-and-running for themselves.

File: resources/views/security/login.twig (hunk 2 above)
Comment (chartjes) [reply resources/views/security/login.twig]: Removed that, must have been left over from my debugging and testing of the feature via manual testing.

File: src/Http/Action/Security/CentralAction.php (hunk 3 above)
Comment (chartjes) [reply src/Http/Action/Security/CentralAction.php]: Removed

File: src/Http/Action/Security/LogInAction.php (hunk 4 above)
Comment (chartjes) [reply src/Http/Action/Security/LogInAction.php]: Removed

File: resources/config/routing.yml (hunk 5 above)
Comment (chartjes) [reply resources/config/routing.yml]: I have made those suggested changes, need to also update the wording on the OpenCFP Central web site to reflect this.

File: resources/config/routing.yml (hunk 6 above)
Comment (chartjes) [reply resources/config/routing.yml]: I have updated all the config files to use the correct authorizeUrl value

File: src/Http/Action/Security/ShowLogInAction.php (hunk 8 above)
Comment (chartjes) [reply src/Http/Action/Security/ShowLogInAction.php]: I don't see a reason to move it. Convince me!

Comment (localheinz) [reply src/Http/Action/Security/ShowLogInAction.php]: None of

```php
$url = $this->urlGenerator->generate('dashboard');

return new HttpFoundation\RedirectResponse($url);
```

will throw an `AuthenticationException`. 

Instead, these two statements constitute the success case on the happy path. By moving them out of the `try` block we make it more clear that they are on that happy path. 

Still, an exception (albeit not an `AuthenticationException`) might be thrown, perhaps when a URL with the `dashboard` identifier could not be generated, but that is clearly out of the scope of handling the authentication.

Does that work?

File: src/Http/Action/Security/SsoAction.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    /** @var string */
+    private $authorizeUrl;
+
+    /**
+     * @var Routing\Generator\UrlGeneratorInterface
+     */
+    private $urlGenerator;
+
+    /** @var string */
+    private $sso;
```
Comment (localheinz) [inline src/Http/Action/Security/SsoAction.php]: Field is still not used, can we remove it?

File: src/Http/Action/Security/SsoAction.php [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * LoginAction constructor.
+     *
+     * @param Services\Authentication                 $authentication
+     * @param Routing\Generator\UrlGeneratorInterface $urlGenerator
+     * @param int                                     $clientId
+     * @param string                                  $redirectUri
+     * @param string                                  $authorizeUrl
+     */
```
Comment (localheinz) [inline src/Http/Action/Security/SsoAction.php]: What do you think about removing the DocBlock, as it only repeats information that is already present from the type declarations of the method arguments? Also, following the rename of the action, the short description has become misleading.

File: src/Http/Action/Security/SsoRedirectAction.php [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+     * @param Sentinel                                $sentinel
+     * @param AccountManagement                       $accounts
+     * @param Routing\Generator\UrlGeneratorInterface $urlGenerator
+     * @param int                                     $clientId
+     * @param string                                  $clientSecret
+     * @param string                                  $redirectUri
+     * @param string                                  $resourceUri
+     * @param string                                  $tokenUrl
+     * @param Client                                  $httpClient
+     */
```
Comment (localheinz) [inline src/Http/Action/Security/SsoRedirectAction.php]: See comment above about suggesting to remove the DocBlock.

File: src/Http/Action/Security/SsoRedirectAction.php [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+    public function __construct(
+        Sentinel $sentinel,
+        AccountManagement $accounts,
+        Routing\Generator\UrlGeneratorInterface $urlGenerator,
+        int $clientId,
+        string $clientSecret,
+        string $redirectUri,
+        string $resourceUri,
+        string $tokenUrl,
```
Comment (localheinz) [inline src/Http/Action/Security/SsoRedirectAction.php]: Not necessary for this PR, but maybe it makes sense to extract (and subsequently, inject here) a configuration object instead?

File: src/Http/Action/Security/SsoRedirectAction.php [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+    /**
+     * @param HttpFoundation\Request $request
+     *
+     * @throws \OpenCFP\Infrastructure\Auth\UserExistsException
+     * @throws \Symfony\Component\Routing\Exception\RouteNotFoundException
+     * @throws \Symfony\Component\Routing\Exception\MissingMandatoryParametersException
+     * @throws \Symfony\Component\Routing\Exception\InvalidParameterException
+     * @throws \OpenCFP\Infrastructure\Auth\UserNotFoundException
+     * @throws \InvalidArgumentException
```
Comment (localheinz) [inline src/Http/Action/Security/SsoRedirectAction.php]: Not sure if any of these exceptions are actually thrown (and not caught) anywhere in the scope of this method, can you check?

File: src/Http/Action/Security/SsoRedirectAction.php [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+        }
+
+        /**
+         * What do we do with the access token?
+         * Check if the user exists
+         * -> if they don't, generate a random password and then create and activate an account
+         * manually log them in
+         * redirect them to their dashboard
+         */
+        $details = \json_decode((string) $response->getBody(), true);
```
Comment (localheinz) [inline src/Http/Action/Security/SsoRedirectAction.php]: Should we check here whether the response body actually contained valid JSON, before moving on and assuming it does?

File: src/Http/Action/Security/SsoRedirectAction.php [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+
+        $userResponse = $this->httpClient->get($this->resourceUri, [
+            'headers' => [
+                'Accept'        => 'application/json',
+                'Authorization' => 'Bearer ' . $details['access_token'],
```
Comment (localheinz) [inline src/Http/Action/Security/SsoRedirectAction.php]: Similarly, should we verify whether the response body actually contains an `access_token` field?

File: src/Http/Action/Security/SsoRedirectAction.php [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+            ],
+        ]);
+        $userDetails = \json_decode((string) $userResponse->getBody(), true);
```
Comment (localheinz) [inline src/Http/Action/Security/SsoRedirectAction.php]: Again, should we check whether the response body actually contains valid JSON?

File: src/Http/Action/Security/SsoRedirectAction.php [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+
+        try {
+            /** @var SentinelUser $user */
+            $user = $this->accounts->findByLogin($userDetails['email']);
```
Comment (localheinz) [inline src/Http/Action/Security/SsoRedirectAction.php]: Should we also verify here whether the response body actually contains an `email` field, and that it is a `string`?

**Comment by localheinz** [comment]
Thank you, @chartjes!

**Pull request merged**